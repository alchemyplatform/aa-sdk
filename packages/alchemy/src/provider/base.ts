import {
  SmartAccountProvider,
  createPublicErc4337Client,
  deepHexlify,
  type BatchUserOperationCallData,
  type UserOperationCallData,
  type UserOperationOverrides,
} from "@alchemy/aa-core";
import { Alchemy } from "alchemy-sdk";
import { type HttpTransport } from "viem";
import { SupportedChains } from "../chains.js";
import { getDefaultUserOperationFeeOptions } from "../defaults.js";
import type { ClientWithAlchemyMethods } from "../middleware/client.js";
import { withAlchemyGasFeeEstimator } from "../middleware/gas-fees.js";
import {
  withAlchemyGasManager,
  type AlchemyGasManagerConfig,
} from "../middleware/gas-manager.js";
import { withAlchemyUserOpSimulation } from "../middleware/simulate-uo.js";
import {
  AlchemyProviderConfigSchema,
  AlchemySdkClientSchema,
} from "../schema.js";
import type { AlchemyProviderConfig } from "../type.js";

export class AlchemyProvider extends SmartAccountProvider<HttpTransport> {
  private rpcUrl: string;

  constructor(config_: AlchemyProviderConfig) {
    const config = AlchemyProviderConfigSchema.parse(config_);

    const { chain, entryPointAddress, opts, ...connectionConfig } = config;
    const _chain =
      typeof chain === "number" ? SupportedChains.get(chain) : chain;
    if (!_chain || !_chain.rpcUrls["alchemy"]) {
      throw new Error(`AlchemyProvider: chain (${chain}) not supported`);
    }

    const rpcUrl =
      connectionConfig.rpcUrl == null
        ? `${_chain.rpcUrls.alchemy.http[0]}/${connectionConfig.apiKey ?? ""}`
        : connectionConfig.rpcUrl;

    const client = createPublicErc4337Client({
      chain: _chain,
      rpcUrl,
      ...(connectionConfig.jwt != null && {
        fetchOptions: {
          headers: {
            Authorization: `Bearer ${connectionConfig.jwt}`,
          },
        },
      }),
    });

    const feeOptions =
      config.opts?.feeOptions ?? getDefaultUserOperationFeeOptions(_chain);

    super({
      rpcProvider: client,
      entryPointAddress,
      chain: _chain,
      opts: { ...opts, feeOptions },
    });

    this.rpcUrl = rpcUrl;

    withAlchemyGasFeeEstimator(this);
  }

  simulateUserOperationAssetChanges = async (
    data: UserOperationCallData | BatchUserOperationCallData,
    overrides?: UserOperationOverrides
  ) => {
    const uoStruct = deepHexlify(
      await this.buildUserOperation(data, overrides)
    );

    return (this.rpcClient as ClientWithAlchemyMethods).request({
      method: "alchemy_simulateUserOperationAssetChanges",
      params: [uoStruct, this.getEntryPointAddress()],
    });
  };

  /**
   * This methods adds the Alchemy Gas Manager middleware to the provider.
   *
   * @param config - the Alchemy Gas Manager configuration
   * @param delegateGasEstimation - whether to delegate the gas estimation entirely to bundler
   *
   * Note that with delegateGasEstimation set to true, the gas estimation middleware will be no-op'd
   * and fee options set upon provider init will be ignored.
   *
   * @returns {AlchemyProvider} - a new AlchemyProvider with the Gas Manager middleware
   */
  withAlchemyGasManager(
    config: AlchemyGasManagerConfig,
    delegateGasEstimation: boolean = true
  ): AlchemyProvider {
    if (!this.isConnected()) {
      throw new Error(
        "AlchemyProvider: account is not set, did you call `connect` first?"
      );
    }

    return withAlchemyGasManager(this, config, {
      estimateGas: delegateGasEstimation,
    });
  }

  /**
   * This method adds the Alchemy UserOperation Simulation middleware to the provider.
   *
   * @returns {AlchemyProvider} - a new AlchemyProvider with UserOperation Simulation middleware
   */
  withAlchemyUserOpSimulation(): this {
    if (!this.isConnected()) {
      throw new Error(
        "AlchemyProvider: account is not set, did you call `connect` first?"
      );
    }

    return withAlchemyUserOpSimulation(this);
  }

  /**
   * This method adds Alchemy Enhanced APIs to the provider, via an optional dependency on
   * `alchemy-sdk`.
   * @see: https://github.com/alchemyplatform/alchemy-sdk-js
   *
   * The Alchemy SDK client must be configured with the same API key and network as the
   * AlchemyProvider. This method validates such at runtime.
   *
   * Additionally, since the Alchemy SDK client does not yet support JWT authentication,
   * AlchemyProvider initialized with JWTs cannot use this method. They must be initialized with an
   * API key or RPC URL.
   *
   * There is an open issue on the Alchemy SDK repo to add JWT support in the meantime.
   * @see: https://github.com/alchemyplatform/alchemy-sdk-js/issues/386
   *
   * @param alchemy - an initialized Alchemy SDK client
   * @returns - a new AlchemyProvider extended with Alchemy SDK client methods
   */
  withAlchemyEnhancedApis(alchemy: Alchemy): this & Alchemy {
    AlchemySdkClientSchema.parse(alchemy);

    if (alchemy.config.url && alchemy.config.url !== this.rpcUrl) {
      throw new Error(
        "Alchemy SDK client JSON-RPC URL must match AlchemyProvider JSON-RPC URL"
      );
    }

    const alchemyUrl = `https://${alchemy.config.network}.g.alchemy.com/v2/${alchemy.config.apiKey}`;
    if (alchemyUrl !== this.rpcUrl) {
      throw new Error(
        "Alchemy SDK client JSON-RPC URL must match AlchemyProvider JSON-RPC URL"
      );
    }

    return this.extend(() => {
      return {
        core: alchemy.core,
        nft: alchemy.nft,
        transact: alchemy.transact,
        debug: alchemy.debug,
        ws: alchemy.ws,
        notify: alchemy.notify,
        config: alchemy.config,
      };
    });
  }
}
