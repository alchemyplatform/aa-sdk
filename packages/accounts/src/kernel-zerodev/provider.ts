import { SmartAccountProvider } from "@alchemy/aa-core";
import type { Hash, HttpTransport } from "viem";
import { KernelSmartContractAccount } from "./account.js";

export class KernelAccountProvider extends SmartAccountProvider<HttpTransport> {
  signMessage = async (msg: string | Uint8Array): Promise<Hash> => {
    if (!this.account) {
      throw new Error("account not connected!");
    }
    return (this.account as KernelSmartContractAccount).signWithEip6492(msg);
  };
}
