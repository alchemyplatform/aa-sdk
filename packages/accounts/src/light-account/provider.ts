import { SmartAccountProvider } from "@alchemy/aa-core";
import { LightSmartContractAccount } from "./account.js";
import { LightAccountProviderConfigSchema } from "./schema.js";
import type { LightAccountProviderConfig } from "./types.js";
import { getDefaultLightAccountFactoryAddress } from "./utils.js";

export const createLightAccountProvider = (
  config_: LightAccountProviderConfig
): SmartAccountProvider & { account: LightSmartContractAccount } => {
  const config = LightAccountProviderConfigSchema.parse(config_);

  return new SmartAccountProvider({
    rpcProvider: config.rpcProvider,
    chain: config.chain,
    entryPointAddress: config.entryPointAddress,
    opts: config.opts,
  }).connect(
    (rpcClient) =>
      new LightSmartContractAccount({
        rpcClient,
        initCode: config.initCode,
        owner: config.owner,
        chain: config.chain,
        entryPointAddress: config.entryPointAddress,
        factoryAddress:
          config.factoryAddress ??
          getDefaultLightAccountFactoryAddress(config.chain),
        accountAddress: config.accountAddress,
      })
  );
};
