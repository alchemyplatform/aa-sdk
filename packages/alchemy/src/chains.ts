import type { Chain } from "viem";
import {
  arbitrum,
  arbitrumGoerli,
  goerli,
  mainnet,
  optimism,
  optimismGoerli,
  polygon,
  polygonMumbai,
  sepolia,
  base,
  baseGoerli,
} from "viem/chains";

export const SupportedChains = new Map<number, Chain>([
  [polygonMumbai.id, polygonMumbai],
  [polygon.id, polygon],
  [mainnet.id, mainnet],
  [sepolia.id, sepolia],
  [goerli.id, goerli],
  [arbitrumGoerli.id, arbitrumGoerli],
  [arbitrum.id, arbitrum],
  [optimism.id, optimism],
  [optimismGoerli.id, optimismGoerli],
  [base.id, base],
  [baseGoerli.id, baseGoerli],
]);
