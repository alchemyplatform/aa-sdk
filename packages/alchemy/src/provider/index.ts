export { AlchemyProvider } from "./base.js";
export { createLightAccountAlchemyProvider } from "./light-account.js";
