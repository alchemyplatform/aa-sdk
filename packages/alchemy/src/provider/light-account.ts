import {
  getDefaultLightAccountFactoryAddress,
  LightSmartContractAccount,
} from "@alchemy/aa-accounts";
import { LightAccountAlchemyProviderConfigSchema } from "../schema.js";
import type { LightAccountAlchemyProviderConfig } from "../type.js";
import { AlchemyProvider } from "./base.js";

/**
 * This method improves the developer experience of connecting a Light Account to an
 * AlchemyProvider via an optional dependency on the `@alchemy/aa-accounts` package.
 * @see: https://github.com/alchemyplatform/aa-sdk/tree/development/packages/accounts
 *
 * @param config_ - the AlchemyProvider configuration with additional pamaeters for Light Account
 * @returns - a new AlchemyProvider connected to a Light Account
 */
export const createLightAccountAlchemyProvider = (
  config_: LightAccountAlchemyProviderConfig
): AlchemyProvider & { account: LightSmartContractAccount } => {
  const config = LightAccountAlchemyProviderConfigSchema.parse(config_);

  return new AlchemyProvider(config).connect(
    (rpcClient) =>
      new LightSmartContractAccount({
        rpcClient,
        initCode: config.initCode,
        owner: config.owner,
        chain: config.chain,
        entryPointAddress: config.entryPointAddress,
        factoryAddress:
          config.factoryAddress ??
          getDefaultLightAccountFactoryAddress(config.chain),
        accountAddress: config.accountAddress,
      })
  );
};
