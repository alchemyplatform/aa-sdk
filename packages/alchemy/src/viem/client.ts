import { createPublicErc4337Client, type Prettify } from "@alchemy/aa-core";
import {
  createSmartAccountClientFromExisting,
  type SmartAccountClient,
  type SmartContractAccount,
} from "@alchemy/aa-core/viem";
import { type Chain, type Transport } from "viem";
import { getDefaultUserOperationFeeOptions } from "../defaults.js";
import type { AlchemyGasManagerConfig } from "../middleware/gas-manager.js";
import { AlchemyProviderConfigSchema } from "../schema.js";
import type { AlchemyProviderConfig } from "../type";
import { alchemyFeeEstimator } from "./middleware/feeEstimator.js";
import { alchemyGasManagerMiddleware } from "./middleware/gasManager.js";
import { alchemyUserOperationSimulator } from "./middleware/userOperationSimulator.js";

export type AlchemySmartAccountClientConfig<
  account extends SmartContractAccount | undefined =
    | SmartContractAccount
    | undefined
> = {
  account?: account;
  useSimulation?: boolean;
  // TODO: this is missing the current gas manager fallback stuff
  gasManagerConfig?: AlchemyGasManagerConfig;
} & AlchemyProviderConfig;

export type AlchemySmartAccountClient<
  transport extends Transport = Transport,
  chain extends Chain = Chain,
  account extends SmartContractAccount | undefined =
    | SmartContractAccount
    | undefined
> = Prettify<SmartAccountClient<transport, chain, account>>;

export function createAlchemySmartAccountClient<
  TTransport extends Transport = Transport,
  TChain extends Chain = Chain,
  TAccount extends SmartContractAccount | undefined =
    | SmartContractAccount
    | undefined
>({
  account,
  gasManagerConfig,
  useSimulation,
  ...config_
}: AlchemySmartAccountClientConfig<TAccount>): AlchemySmartAccountClient<
  TTransport,
  TChain,
  TAccount
>;

// TODO: this is missing the additional logic for the gas manager fallback stuff
// TODO: this is missing the simulation action
// TODO: this is missing the extend with alchemy sdk stuff
// TODO: we should consider allowing this client to be created with a PublicErc4337Client that way one public client can be shared between
// the account and this one. We could probably export an AlchemyRpcClient from this package which extends PublicErc4337Client
export function createAlchemySmartAccountClient<
  TAccount extends SmartContractAccount | undefined =
    | SmartContractAccount
    | undefined
>({
  account,
  gasManagerConfig,
  useSimulation,
  ...config_
}: AlchemySmartAccountClientConfig<TAccount>): AlchemySmartAccountClient {
  const config = AlchemyProviderConfigSchema.parse(config_);
  const { chain, opts, ...connectionConfig } = config;

  const rpcUrl =
    connectionConfig.rpcUrl == null
      ? `${chain.rpcUrls.alchemy.http[0]}/${connectionConfig.apiKey ?? ""}`
      : connectionConfig.rpcUrl;

  const client = createPublicErc4337Client({
    chain: chain,
    rpcUrl,
    ...(connectionConfig.jwt != null && {
      fetchOptions: {
        headers: {
          Authorization: `Bearer ${connectionConfig.jwt}`,
        },
      },
    }),
  });

  const feeOptions =
    config.opts?.feeOptions ?? getDefaultUserOperationFeeOptions(chain);

  return createSmartAccountClientFromExisting({
    account,
    client,
    opts: {
      ...opts,
      feeOptions,
    },
    feeEstimator: alchemyFeeEstimator(client),
    userOperationSimulator: useSimulation
      ? alchemyUserOperationSimulator(client)
      : undefined,
    ...(gasManagerConfig
      ? alchemyGasManagerMiddleware(client, gasManagerConfig)
      : {}),
  });
}
