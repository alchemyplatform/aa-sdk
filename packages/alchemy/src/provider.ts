import {
  SmartAccountProvider,
  createPublicErc4337Client,
  deepHexlify,
  resolveProperties,
  type AccountMiddlewareFn,
  type BatchUserOperationCallData,
  type UserOperationCallData,
  type UserOperationOverrides,
} from "@alchemy/aa-core";
import { Alchemy } from "alchemy-sdk";
import { type HttpTransport } from "viem";
import {
  arbitrum,
  arbitrumGoerli,
  optimism,
  optimismGoerli,
} from "viem/chains";
import { SupportedChains } from "./chains.js";
import type { ClientWithAlchemyMethods } from "./middleware/client.js";
import { withAlchemyGasFeeEstimator } from "./middleware/gas-fees.js";
import {
  withAlchemyGasManager,
  type AlchemyGasManagerConfig,
} from "./middleware/gas-manager.js";
import { withAlchemyUserOpSimulation } from "./middleware/simulate-uo.js";
import {
  AlchemyProviderConfigSchema,
  AlchemySdkClientSchema,
} from "./schema.js";
import type { AlchemyProviderConfig } from "./type.js";

export class AlchemyProvider extends SmartAccountProvider<HttpTransport> {
  private pvgBuffer: bigint;
  private feeOptsSet: boolean;
  private rpcUrl: string;

  constructor(config: AlchemyProviderConfig) {
    AlchemyProviderConfigSchema.parse(config);

    const { chain, entryPointAddress, opts, feeOpts, ...connectionConfig } =
      config;
    const _chain =
      typeof chain === "number" ? SupportedChains.get(chain) : chain;
    if (!_chain || !_chain.rpcUrls["alchemy"]) {
      throw new Error(`AlchemyProvider: chain (${chain}) not supported`);
    }

    const rpcUrl =
      connectionConfig.rpcUrl == null
        ? `${_chain.rpcUrls.alchemy.http[0]}/${connectionConfig.apiKey ?? ""}`
        : connectionConfig.rpcUrl;

    const client = createPublicErc4337Client({
      chain: _chain,
      rpcUrl,
      ...(connectionConfig.jwt != null && {
        fetchOptions: {
          headers: {
            Authorization: `Bearer ${connectionConfig.jwt}`,
          },
        },
      }),
    });

    super({
      rpcProvider: client,
      entryPointAddress,
      chain: _chain,
      opts,
    });

    withAlchemyGasFeeEstimator(
      this,
      feeOpts?.baseFeeBufferPercent ?? 50n,
      feeOpts?.maxPriorityFeeBufferPercent ?? 5n
    );

    if (feeOpts?.preVerificationGasBufferPercent) {
      this.pvgBuffer = feeOpts?.preVerificationGasBufferPercent;
    } else if (
      new Set<number>([
        arbitrum.id,
        arbitrumGoerli.id,
        optimism.id,
        optimismGoerli.id,
      ]).has(this.chain.id)
    ) {
      this.pvgBuffer = 5n;
    } else {
      this.pvgBuffer = 0n;
    }

    this.feeOptsSet = !!feeOpts;
    this.rpcUrl = rpcUrl;
  }

  override gasEstimator: AccountMiddlewareFn = async (struct) => {
    const request = deepHexlify(await resolveProperties(struct));
    const estimates = await this.rpcClient.estimateUserOperationGas(
      request,
      this.getEntryPointAddress()
    );
    estimates.preVerificationGas =
      (BigInt(estimates.preVerificationGas) * (100n + this.pvgBuffer)) / 100n;

    return {
      ...struct,
      ...estimates,
    };
  };

  simulateUserOperationAssetChanges = async (
    data: UserOperationCallData | BatchUserOperationCallData,
    overrides?: UserOperationOverrides
  ) => {
    const uoStruct = deepHexlify(
      await this.buildUserOperation(data, overrides)
    );

    return (this.rpcClient as ClientWithAlchemyMethods).request({
      method: "alchemy_simulateUserOperationAssetChanges",
      params: [uoStruct, this.getEntryPointAddress()],
    });
  };

  /**
   * This methods adds the Alchemy Gas Manager middleware to the provider.
   *
   * @param config - the Alchemy Gas Manager configuration
   * @returns {AlchemyProvider} - a new AlchemyProvider with the Gas Manager middleware
   */
  withAlchemyGasManager(config: AlchemyGasManagerConfig): AlchemyProvider {
    if (!this.isConnected()) {
      throw new Error(
        "AlchemyProvider: account is not set, did you call `connect` first?"
      );
    }

    return withAlchemyGasManager(this, config, !this.feeOptsSet);
  }

  /**
   * This methods adds the Alchemy UserOperation Simulation middleware to the provider.
   *
   * @returns {AlchemyProvider} - a new AlchemyProvider with the UserOperation Simulation middleware
   */
  withAlchemyUserOpSimulation(): this {
    if (!this.isConnected()) {
      throw new Error(
        "AlchemyProvider: account is not set, did you call `connect` first?"
      );
    }

    return withAlchemyUserOpSimulation(this);
  }

  /**
   * This methods adds Alchemy Enhanced APIs to the provider, via an optional dependency on `alchemy-sdk`.
   * @see: https://github.com/alchemyplatform/alchemy-sdk-js
   *
   * The Alchemy SDK client must be configured with the same API key and network as the AlchemyProvider.
   * This method validates such at runtime.
   *
   * Additionally, since the Alchemy SDK client does not yet support JWT authentication, AlchemyProviders initialized with JWTs cannot use this method.
   * They must be initialized with an API key or RPC URL.
   * There is an open issue on the Alchemy SDK repo to add JWT support in the meantime.
   * @see: https://github.com/alchemyplatform/alchemy-sdk-js/issues/386
   *
   * @param alchemy - an initialized Alchemy SDK client
   * @returns - a new AlchemyProvider extended with Alchemy SDK client methods
   */
  withAlchemyEnhancedApis(alchemy: Alchemy): this & Alchemy {
    AlchemySdkClientSchema.parse(alchemy);

    if (alchemy.config.url && alchemy.config.url !== this.rpcUrl) {
      throw new Error(
        "Alchemy SDK client JSON-RPC URL must match AlchemyProvider JSON-RPC URL"
      );
    }

    const alchemyUrl = `https://${alchemy.config.network}.g.alchemy.com/v2/${alchemy.config.apiKey}`;
    if (alchemyUrl !== this.rpcUrl) {
      throw new Error(
        "Alchemy SDK client JSON-RPC URL must match AlchemyProvider JSON-RPC URL"
      );
    }

    return this.extend(() => {
      return {
        core: alchemy.core,
        nft: alchemy.nft,
        transact: alchemy.transact,
        debug: alchemy.debug,
        ws: alchemy.ws,
        notify: alchemy.notify,
        config: alchemy.config,
      };
    });
  }
}
