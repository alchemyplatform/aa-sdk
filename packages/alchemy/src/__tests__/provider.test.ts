import * as AACoreModule from "@alchemy/aa-core";
import {
  SimpleSmartContractAccount,
  type BatchUserOperationCallData,
  type SmartAccountSigner,
} from "@alchemy/aa-core";
import { toHex, type Address, type Chain } from "viem";
import { mnemonicToAccount } from "viem/accounts";
import { polygonMumbai } from "viem/chains";
import { AlchemyProvider } from "../provider.js";

describe("Alchemy Provider Tests", () => {
  const dummyMnemonic =
    "test test test test test test test test test test test test";
  const ownerAccount = mnemonicToAccount(dummyMnemonic);
  const owner: SmartAccountSigner = {
    signMessage: async (msg) =>
      ownerAccount.signMessage({
        message: { raw: toHex(msg) },
      }),
    signTypedData: async () => "0xHash",
    getAddress: async () => ownerAccount.address,
    signerType: "e2e-test",
  };
  const chain = polygonMumbai;

  it("should have a JWT propety", async () => {
    const spy = vi.spyOn(AACoreModule, "createPublicErc4337Client");
    givenConnectedProvider({ owner, chain });
    expect(spy.mock.calls[0][0].fetchOptions).toMatchInlineSnapshot(`
      {
        "headers": {
          "Authorization": "Bearer test",
        },
      }
    `);
  });

  it("should correctly sign the message", async () => {
    const signer = givenConnectedProvider({ owner, chain });
    expect(
      // TODO: expose sign message on the provider too
      await signer.account.signMessage(
        "0xa70d0af2ebb03a44dcd0714a8724f622e3ab876d0aa312f0ee04823285d6fb1b"
      )
    ).toBe(
      "0x33b1b0d34ba3252cd8abac8147dc08a6e14a6319462456a34468dd5713e38dda3a43988460011af94b30fa3efefcf9d0da7d7522e06b7bd8bff3b65be4aee5b31c"
    );
  });

  it("should correctly encode batch transaction data", async () => {
    const signer = givenConnectedProvider({ owner, chain });
    const account = signer.account;
    const data = [
      {
        target: "0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef",
        data: "0xdeadbeef",
      },
      {
        target: "0x8ba1f109551bd432803012645ac136ddd64dba72",
        data: "0xcafebabe",
      },
    ] satisfies BatchUserOperationCallData;

    expect(await account.encodeBatchExecute(data)).toMatchInlineSnapshot(
      '"0x18dfb3c7000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000002000000000000000000000000deadbeefdeadbeefdeadbeefdeadbeefdeadbeef0000000000000000000000008ba1f109551bd432803012645ac136ddd64dba720000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000004deadbeef000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004cafebabe00000000000000000000000000000000000000000000000000000000"'
    );
  });
});

const givenConnectedProvider = ({
  owner,
  chain,
}: {
  owner: SmartAccountSigner;
  chain: Chain;
}) => {
  const dummyEntryPointAddress =
    "0x1234567890123456789012345678901234567890" as Address;

  return new AlchemyProvider({
    rpcUrl: "https://eth-mainnet.g.alchemy.com/v2",
    jwt: "test",
    chain,
  }).connect((provider) => {
    const account = new SimpleSmartContractAccount({
      entryPointAddress: dummyEntryPointAddress,
      chain,
      owner,
      factoryAddress: "0xSIMPLE_ACCOUNT_FACTORY_ADDRESS",
      rpcClient: provider,
    });

    account.getAddress = vi.fn(
      async () => "0xb856DBD4fA1A79a46D426f537455e7d3E79ab7c4"
    );

    return account;
  });
};
