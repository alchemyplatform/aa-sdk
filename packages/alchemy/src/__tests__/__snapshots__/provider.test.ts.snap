// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alchemy Provider Tests > should correctly do runtime validation when connection config is invalid 1`] = `
"[
  {
    \\"code\\": \\"invalid_union\\",
    \\"unionErrors\\": [
      {
        \\"issues\\": [
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"never\\",
            \\"received\\": \\"number\\",
            \\"path\\": [
              \\"rpcUrl\\"
            ],
            \\"message\\": \\"Expected never, received number\\"
          },
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"string\\",
            \\"received\\": \\"undefined\\",
            \\"path\\": [
              \\"apiKey\\"
            ],
            \\"message\\": \\"Required\\"
          }
        ],
        \\"name\\": \\"ZodError\\"
      },
      {
        \\"issues\\": [
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"never\\",
            \\"received\\": \\"number\\",
            \\"path\\": [
              \\"rpcUrl\\"
            ],
            \\"message\\": \\"Expected never, received number\\"
          },
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"string\\",
            \\"received\\": \\"undefined\\",
            \\"path\\": [
              \\"jwt\\"
            ],
            \\"message\\": \\"Required\\"
          }
        ],
        \\"name\\": \\"ZodError\\"
      },
      {
        \\"issues\\": [
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"string\\",
            \\"received\\": \\"number\\",
            \\"path\\": [
              \\"rpcUrl\\"
            ],
            \\"message\\": \\"Expected string, received number\\"
          }
        ],
        \\"name\\": \\"ZodError\\"
      },
      {
        \\"issues\\": [
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"string\\",
            \\"received\\": \\"number\\",
            \\"path\\": [
              \\"rpcUrl\\"
            ],
            \\"message\\": \\"Expected string, received number\\"
          },
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"string\\",
            \\"received\\": \\"undefined\\",
            \\"path\\": [
              \\"jwt\\"
            ],
            \\"message\\": \\"Required\\"
          }
        ],
        \\"name\\": \\"ZodError\\"
      }
    ],
    \\"path\\": [],
    \\"message\\": \\"Invalid input\\"
  }
]"
`;

exports[`Alchemy Provider Tests > should have enhanced api properties extended from the Alchemy SDK 1`] = `
Alchemy {
  "config": AlchemyConfig {
    "apiKey": "test",
    "authToken": undefined,
    "batchRequests": false,
    "maxRetries": 5,
    "network": "polygon-mumbai",
    "requestTimeout": 0,
    "url": undefined,
  },
  "core": CoreNamespace {
    "config": AlchemyConfig {
      "apiKey": "test",
      "authToken": undefined,
      "batchRequests": false,
      "maxRetries": 5,
      "network": "polygon-mumbai",
      "requestTimeout": 0,
      "url": undefined,
    },
  },
  "debug": DebugNamespace {
    "config": AlchemyConfig {
      "apiKey": "test",
      "authToken": undefined,
      "batchRequests": false,
      "maxRetries": 5,
      "network": "polygon-mumbai",
      "requestTimeout": 0,
      "url": undefined,
    },
  },
  "nft": NftNamespace {
    "config": AlchemyConfig {
      "apiKey": "test",
      "authToken": undefined,
      "batchRequests": false,
      "maxRetries": 5,
      "network": "polygon-mumbai",
      "requestTimeout": 0,
      "url": undefined,
    },
  },
  "notify": NotifyNamespace {
    "config": AlchemyConfig {
      "apiKey": "test",
      "authToken": undefined,
      "batchRequests": false,
      "maxRetries": 5,
      "network": "polygon-mumbai",
      "requestTimeout": 0,
      "url": undefined,
    },
  },
  "transact": TransactNamespace {
    "config": AlchemyConfig {
      "apiKey": "test",
      "authToken": undefined,
      "batchRequests": false,
      "maxRetries": 5,
      "network": "polygon-mumbai",
      "requestTimeout": 0,
      "url": undefined,
    },
  },
  "ws": WebSocketNamespace {
    "config": AlchemyConfig {
      "apiKey": "test",
      "authToken": undefined,
      "batchRequests": false,
      "maxRetries": 5,
      "network": "polygon-mumbai",
      "requestTimeout": 0,
      "url": undefined,
    },
  },
}
`;
